src/librustc_mir/interpret/validity.rs

```diff
@@ -578,26 +578,36 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
                     self.path
                 );
             }
-            ty::Array(tys, ..) | ty::Slice(tys)
-                if {
-                    // This optimization applies for types that can hold arbitrary bytes (such as
-                    // integer and floating point types) or for structs or tuples with no fields.
-                    // FIXME(wesleywiser) This logic could be extended further to arbitrary structs
-                    // or tuples made up of integer/floating point types or inhabited ZSTs with no
-                    // padding.
-                    match tys.kind {
-                        ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,
-                        ty::Tuple(tys) if tys.len() == 0 => true,
-                        ty::Adt(adt_def, _)
-                            if adt_def.is_struct() && adt_def.all_fields().next().is_none() =>
-                        {
-                            true
-                        }
-                        _ => false,
+            ty::Array(tys, ..) | ty::Slice(tys) => {
+                // This optimization applies for types that can hold arbitrary bytes (such as
```

Comment:
The comment here is weird now... the optimization comes way later, *after* the default case, and the first optimization is only for ZST, not what this comment says.

Reply:
I'd structure this roughly as follows:
* first optimization: ZST elem type
* second optimization: scalar elem type where all types are valid
* general case

Not sure if you can make the code actually follow this structure, but IMO this makes most sense for the comments.

Reply:
that would mean that the "scalar elem type where all types are valid" optimization is inside one big if-block with a tiny else block for the general case in the end. I tried to avoid that.

Reply:
Yeah I know, that's not great either. Maybe make the `check_bytes` into a separate function?